x/superfluid/keeper/migrate.go

```diff
@@ -155,160 +156,157 @@ func (k Keeper) migrateSuperfluidUnbondingBalancerToConcentrated(ctx sdk.Context
-		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
 	}
 
 	// Create a new lock that is unlocking for the remaining time of the old lock.
 	// Regardless of the previous lock's status, we create a new lock that is unlocking.
+	// This is because locking without superfluid is pointless in the context of concentrated liquidity.
 	positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionUnlocking(ctx, poolIdEntering, sender, exitCoins, remainingLockTime)
 	if err != nil {
-		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
 	}
 
 	// If there are remaining gamm shares, we must re-lock them.
-	remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
-	if !remainingGammShares.IsZero() {
-		// Create a new lock with the remaining gamm shares for the remaining lock time.
-		newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
-		gammLockId = newLock.ID
-		if err != nil {
-			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
-		}
-
+	if remainingSharesLock.ID != 0 {
+		gammLockId = remainingSharesLock.ID
 		// If the gamm lock was unlocking, we begin the unlock from where it left off.
 		if wasUnlocking {
-			_, err := k.lk.BeginForceUnlock(ctx, newLock.ID, newLock.Coins)
+			_, err := k.lk.BeginForceUnlock(ctx, remainingSharesLock.ID, remainingSharesLock.Coins)
 			if err != nil {
-				return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+				return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
 			}
 		}
 	}
 
-	return positionId, amount0, amount1, liquidity, joinTime, gammLockId, concentratedLockId, nil
+	return positionId, amount0, amount1, liquidity, joinTime, gammLockId, concentratedLockId, poolIdLeaving, poolIdEntering, nil
 }
 
-// prepareMigration prepares for the migration of gamm LP tokens from the Balancer pool to the Concentrated pool. It performs the following steps:
+// routeMigration determines the status of the provided lock which is used to determine the method for migration.
+// It also returns the underlying synthetic locks of the provided lock, if any exist.
+func (k Keeper) routeMigration(ctx sdk.Context, sender sdk.AccAddress, lockId uint64, sharesToMigrate sdk.Coin) (synthLocksBeforeMigration []lockuptypes.SyntheticLock, migrationType MigrationType, err error) {
+	synthLocksBeforeMigration = k.lk.GetAllSyntheticLockupsByLockup(ctx, lockId)
+	migrationType = NonSuperfluid
+
+	for _, synthLockBeforeMigration := range synthLocksBeforeMigration {
+		if strings.Contains(synthLockBeforeMigration.SynthDenom, "superbonding") {
+			migrationType = SuperfluidBonded
+		}
+		if strings.Contains(synthLockBeforeMigration.SynthDenom, "superunbonding") {
+			migrationType = SuperfluidUnbonding
+		}
+	}
+
```

Comment:
what happens if there are two synthlocks existing, one being superfluid bonded, one being unbonding?

Also curious if `synthLocksBeforeMigration` is sorted deterministically

Reply:
In my audit, I have this listed as an unknown. In what instance does a single lock have both unbonding and bonded synth denoms? Right now, this just errors and is unsupported. But I want to know if this is a normal occurrence

Reply:
Is it the case that this errors right now? At a brief glance, within this function, it seems that while iterating `synthLocksBeforeMigration`, if the length of `synthLocksBeforeMigration` is more than 1, `migrationType` would be able to continuously change. 

I have checked the code to find out in which case a single lock could have both unbonding and bonded synth denoms, but could not find an answer for that. Regardless, wdyt of adding a check here just to make sure length of `GetAllSyntheticLockupsByLockup` is not more than 1?

Reply:
After DMs, agreed, my bad. Fixed here https://github.com/osmosis-labs/osmosis/pull/5160/commits/478aa100de7c3bea8d57a4514ef2b3a9744d909f